### PR TITLE
test: use @bazel_tools//tools/cpp/runfiles to find runfiles

### PR DESCRIPTION
**Pull request by lizan**
Description:
Proper fix for bazelbuild/bazel#7399 from Envoy side, tests should be aware which workspace its testdata is from. Added parameter to TestEnvironment::runfiles{Path,Directory} to specify `workspace`. Using this ability in compile_time_options CI.

By using `@bazel_tools//tools/cpp/runfiles`, it also allows test binaries to be invoked outside bazel directly, e.g. `bazel-bin/test/exe/main_common_test`, so the GDB wrapper is no longer needed.

Risk Level: Low (test only)
Testing: CI
Docs Changes: N/A
Release Notes: N/A

Signed-off-by: Lizan Zhou <lizan@tetrate.io>

### CODE REVIEW

File: ci/do_ci.sh [hunk 1]
```diff
@@ -203,9 +203,14 @@ elif [[ "$CI_TARGET" == "bazel.compile_time_options" ]]; then
   setup_clang_libcxx_toolchain
   # This doesn't go into CI but is available for developer convenience.
   echo "bazel with different compiletime options build with tests..."
+
+  if [[ "${TEST_TARGETS}" == "//test/..." ]]; then
+    cd "${ENVOY_FILTER_EXAMPLE_SRCDIR}"
+    TEST_TARGETS="@envoy//test/..."
```
Comment (htuch) [inline ci/do_ci.sh]: This seems a bit special casey..

File: STYLE.md [hunk 2]
```diff
@@ -165,7 +165,7 @@ environment. In general, there should be no non-local network access. In additio
 
 * Paths should be constructed using:
   * The methods in [`TestEnvironment`](test/test_common/environment.h) for C++ tests.
-  * With `${TEST_TMPDIR}` (for writable temporary space) or `${TEST_RUNDIR}` for read-only access to
+  * With `${TEST_TMPDIR}` (for writable temporary space) or `${TEST_SRCDIR}` for read-only access to
```
[the remark this answers is not in the document]
Comment (lizan) [reply STYLE.md]: same, genrule generated artifacts is in the TEST_SRCDIR for read-only

File: test/main.cc [hunk 3]
```diff
@@ -22,10 +25,13 @@ int main(int argc, char** argv) {
   Envoy::SignalAction handle_sigs;
 #endif
 
-  Envoy::TestEnvironment::setEnvVar("TEST_RUNDIR",
-                                    (Envoy::TestEnvironment::getCheckedEnvVar("TEST_SRCDIR") + "/" +
-                                     Envoy::TestEnvironment::getCheckedEnvVar("TEST_WORKSPACE")),
-                                    1);
+  std::string error;
+  std::unique_ptr<Runfiles> runfiles(Runfiles::Create(argv[0], &error));
```
Comment (htuch) [inline test/main.cc]: Can you add some comments explaining what this magic does?

Comment (lizan) [reply test/main.cc]: done

File: ci/do_ci.sh (hunk 1 above)
Comment (lizan) [reply ci/do_ci.sh]: This is for the default case (where TEST_TARGETS is not specified), and for CI setup.

**Comment by lizan** [comment]
@htuch now the tests are all passed except those requiring https://github.com/envoyproxy/envoy-filter-example/pull/101.

**Comment by htuch** [comment]
@lizan merge the envoy-filter-example PR.

**Pull request merged**